src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php

```diff
@@ -64,33 +99,116 @@ public function update(Response $response)
+            $maxAge = $this->ageDirectives['max-age'] + $this->age;
+            $response->headers->addCacheControlDirective('max-age', $maxAge);
         }
 
-        if (!$this->cacheable) {
-            $response->headers->set('Cache-Control', 'no-cache, must-revalidate');
+        if ($this->ageDirectives['s-maxage']) {
+            $sMaxage = $this->ageDirectives['s-maxage'] + $this->age;
 
-            return;
+            if ($maxAge !== $sMaxage) {
+                $response->headers->addCacheControlDirective('s-maxage', $sMaxage);
+            }
+        }
+
+        if ($this->ageDirectives['expires']) {
+            $response->setExpires($response->getDate()->modify('+'.$this->ageDirectives['expires'].' seconds'));
         }
+    }
 
-        $this->ttls[] = $response->getTtl();
-        $this->maxAges[] = $response->getMaxAge();
+    /**
+     * RFC2616, Section 13.4.
+     *
+     * @see https://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.4
+     *
+     * @return bool
+     */
+    private function willMakeFinalResponseUncacheable(Response $response)
+    {
+        // RFC2616: A response received with a status code of 200, 203, 300, 301 or 410
+        // MAY be stored by a cache […] unless a cache-control directive prohibits caching.
+        if ($response->headers->hasCacheControlDirective('no-cache')
+            || $response->headers->getCacheControlDirective('no-store')
+        ) {
+            return true;
+        }
 
-        if ($this->isNotCacheableResponseEmbedded) {
-            $response->headers->removeCacheControlDirective('s-maxage');
-        } elseif (null !== $maxAge = min($this->maxAges)) {
-            $response->setSharedMaxAge($maxAge);
-            $response->headers->set('Age', $maxAge - min($this->ttls));
+        // Last-Modified and Etag headers cannot be merged, they render the response uncacheable
+        // by default (except if the response also has max-age etc.).
+        if (\in_array($response->getStatusCode(), array(200, 203, 300, 301, 410))
+            && null === $response->getLastModified()
+            && null === $response->getEtag()
+        ) {
+            return false;
+        }
+
+        // RFC2616: A response received with any other status code (e.g. status codes 302 and 307)
+        // MUST NOT be returned in a reply to a subsequent request unless there are
+        // cache-control directives or another header(s) that explicitly allow it.
+        $cacheControl = array('max-age', 's-maxage', 'must-revalidate', 'proxy-revalidate', 'public', 'private');
+        foreach ($cacheControl as $key) {
+            if ($response->headers->hasCacheControlDirective($key)) {
+                return false;
+            }
+        }
+
+        if ($response->headers->has('Expires')) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Stores the cache directive relative to the age of the response.
```

Comment:
can you please explain how those things are handled? the line above essentially repeats the method name. instead, it would be good to be clear how instructions and the current age are accumulated into the age directives of the master response. (do we always end up with an age of 0, but deduct the current age of each fragment from its max-age and s-maxage?)

Reply:
Good point. Would something like this make sense to you?

> Calculates the current value of max-age/s-maxage/expires by subtracting age, because the response might have been stored in the cache a while ago, so it does not expire X seconds from now, but X seconds from when it was created. It then stores the value if it is lower than any previous responses.

Reply:
yep. maybe like this:
```
/**
 * Lower max-age/s-maxage/expires for the whole response, if needed.
 *
 * The fragment might have been stored in cache a while ago. To keep things manageable, 
 * we substract the age so that all values are normalized for an age of 0.
 *
 * If the value is lower than the currently stored value, we update the value, to keep a rolling
 * minimal value of each instruction.
 */
```

Reply:
Updated in 13febfb42effec9638d55db20434b3645f6314a2